SHA improvements

petermcs opened the pull request:
This pull request provides code size and speed improvements for the SHA256 code.

petermcs committed:
Rabbit 6000 optimization

Used R6K swap instruction for speed improvement on R6K systems. Also used de for increment as this is quicker than pointer register + offset for updating.

petermcs committed:
Refactor of SHA256 code to reduce size and speed up

Major refactor of sha256_process() and sub functions to dramatically decrease code size and increase performance.
Also includes R6K optimizations for further performance improvements.

On R4K systems the previous code was 8,998 bytes in length and took 17,348mS to hash 1,000,000 bytes in Samples\Crypto\sha.c. The revised code is 4,332 bytes in length and completes 1,000.000 bytes in 11,472mS.
For R6K, the previous code was 8,998 bytes in length and took 6,606mS. The revised code is 4256 bytes in length and completes 1,000,000 bytes in 3,570mS.
To achieve the improvements, the R and P macros have been replaced with assembler macros and the code heavily optimized to reduce memory load/stores and stack push/pull operations for both platforms.
For the R6K, additional optimizations include use of new addressing modes for maths operations to reduce loading of BCDE, alignment statements to reduce cycle counts, 32bit constant loads via pointer registers and use of altd/alts for saving temporary values in  alt pointer regs.

tomlogic commented on Lib/Rabbit4000/Crypto/SHA2.LIB:
```diff
@@ -145,14 +145,16 @@ void sha256_add(sha256_context __far *ctx, const uint8_t __far *input,
-_sha2_debug
+root _sha2_debug
+// result in jkhl and not bcde
+// all arguments in registers and no stack use
```
Since `_sha256_F0_S2()` is no longer callable from C (due to not using Dynamic C's ABI for function calls), you should remove the C prototypes and have the function entirely contained in a `#asm` block, using `_sha256_F0_S2::` as the opening label (ala `_sha256_S0`).

Be sure to document the interface to the function (e.g., pass parameters x, y, and z in registers PX, PY, and PZ; return value in JKHL) so it's easy to validate the implementation and calling code.

tomlogic commented on Lib/Rabbit4000/Crypto/SHA1.LIB:
```diff
@@ -953,26 +953,33 @@ __xmem void sha_copy_and_swap(void __far* dest, const void __far* src, word lwco
+   swap		jkhl
+   ld		(px),jkhl
+#else
+   ld		hl,(py+2)
```
I think you could improve the non-6K code here, based on the technique used in `intel()`:

```
	ex		de,hl
	ld		e,b
	ld		d,c
	ld		b,l
	ld		c,h
```

I'm going to push out a separate commit to make use of the `swap` opcode in `intel()` and `intel16()` on the 6000 platform.

petermcs replied:
Unfortunately, the code from the intel function uses bc and de which we are already using for the loop in this case.
This is probably the best we can do:
   ld		jkhl,(py)
   ld		a,h
   ld		h,L
   ld		L,a
   ex       jk, hl
   ld		a,h
   ld		h,L
   ld		L,a
   ld		(px),jkhl

petermcs replied:
The lack of instructions for accessing j and k or jk is a big pain for these type of operations!

petermcs committed:
Improve copy and swap for non R6K by reducing load/stores

petermcs committed:
Further performance improvements, mainly for R6K including use of SHAF1 and SHAF2 instructions.
Claeaned out some unused macros and commented asm function parameters and returns.
Added option to speed up R6K at expense of code size increase by inlining functions.
BTW just making sha256_add() root shaved 400mS off the R6K time.

R6K code size now 3887 and 1,000,000 bytes in 2,685mS
Alternative R6K code size now 6801 and 1,000,000 bytes in 2,488mS
R4K Code size now 4332 and 1,000,000 bytes in 11,214mS
